server/src/main/java/org/elasticsearch/index/engine/MergeTracking.java

```diff
@@ -50,13 +53,26 @@ public Set<OnGoingMerge> onGoingMerges() {
         return readOnlyOnGoingMerges;
     }
 
+    public Set<OnGoingMerge> queuedMerges() {
+        return readOnlyQueuedMerges;
+    }
+
+    public void markMergeQueued(OnGoingMerge merge) {
```

Comment:
I do see MergeTracking being used in stateless but also in stateful (InternalEngine). So I wonder why we do not call this function also for the stateful merges? An oversight? Or is it because merges cannot be queued in stateful and they just immediately execute?

If we do not mark queued merges in core ES, at least we should document it in the javadoc of `markMergeQueued`, `hasQueuedOrOnGoingMerges()` and `hasQueuedMerges()`, so at least people are not misled to use the functions in stateful without understanding that queued merges are not there.

Reply:
> Or is it because merges cannot be queued in stateful and they just immediately execute?

In stateful merges get executed immediately in their own new thread (https://github.com/elastic/elasticsearch/blob/a8de5545d82120ce316d12e4372e145f1d5a2fa4/server/src/main/java/org/elasticsearch/index/engine/ElasticsearchConcurrentMergeScheduler.java#L136-L143).

> If we do not mark queued merges in core ES, at least we should document it in the javadoc of markMergeQueued, hasQueuedOrOnGoingMerges() and hasQueuedMerges(), so at least people are not misled to use the functions in stateful without understanding that queued merges are not there.

Sure, I'll add a javadoc and mention that it's implementation dependant 👍

Reply:
Oh OK thanks for explaining! Yea javadoc to maybe explain what you just said would be nice so that people know in stateful there are no queued merges.